optaplanner-core/src/main/java/org/optaplanner/core/impl/io/jaxb/GenericJaxbIO.java

```diff
@@ -129,119 +155,141 @@ public T readAndValidate(Reader reader, Schema schema) {
-        } catch (ParserConfigurationException | SAXException e) {
-            final String errMessage = String.format(ERR_MSG_READ_OVERRIDE_NAMESPACE, rootClass.getName(),
+            return (T) createUnmarshaller().unmarshal(translatedDocument);
+        } catch (JAXBException e) {
+            final String errorMessage = String.format(ERR_MSG_READ_OVERRIDE_NAMESPACE, rootClass.getName(),
                     Arrays.toString(elementNamespaceOverrides));
-            throw new OptaPlannerXmlSerializationException(errMessage, e);
+            throw new OptaPlannerXmlSerializationException(errorMessage, e);
         }
-        XMLReader xmlReader;
+    }
+
+    public Document parseXml(Reader reader) {
+        Reader nonNullReader = Objects.requireNonNull(reader);
+        DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
+        documentBuilderFactory.setAttribute(XMLConstants.ACCESS_EXTERNAL_DTD, "");
+        documentBuilderFactory.setAttribute(XMLConstants.ACCESS_EXTERNAL_SCHEMA, "");
+        documentBuilderFactory.setNamespaceAware(true);
+        DocumentBuilder builder;
         try {
-            xmlReader = saxParser.getXMLReader();
-        } catch (SAXException e) {
-            final String errMessage = String.format(ERR_MSG_READ_OVERRIDE_NAMESPACE, rootClass.getName(),
-                    Arrays.toString(elementNamespaceOverrides));
-            throw new OptaPlannerXmlSerializationException(errMessage, e);
+            builder = documentBuilderFactory.newDocumentBuilder();
+        } catch (ParserConfigurationException e) {
+            String errorMessage = String.format("Failed to create a %s instance to parse an XML for a root class (%s).",
+                    DocumentBuilder.class.getSimpleName(), rootClass.getName());
+            throw new OptaPlannerXmlSerializationException(errorMessage, e);
         }
 
-        XMLFilter namespaceOverridingXmlFilter = new NamespaceOverridingXmlFilter(xmlReader, elementNamespaceOverrides);
-        namespaceOverridingXmlFilter.setParent(xmlReader);
-
-        // Use UnmarshallerHandler as a content handler for the XML filter.
-        Unmarshaller unmarshaller = createUnmarshaller();
-        UnmarshallerHandler unmarshallerHandler = unmarshaller.getUnmarshallerHandler();
-        namespaceOverridingXmlFilter.setContentHandler(unmarshallerHandler);
-
-        InputSource xmlInputSource = new InputSource(reader);
         try {
-            // Parse the XML to feed its content into the UnmarshallerHandler.
-            namespaceOverridingXmlFilter.parse(xmlInputSource);
-        } catch (IOException | SAXException e) {
-            final String errMessage = String.format(ERR_MSG_READ_OVERRIDE_NAMESPACE, rootClass.getName(),
-                    Arrays.toString(elementNamespaceOverrides));
-            throw new OptaPlannerXmlSerializationException(errMessage, e);
+            return builder.parse(new InputSource(nonNullReader));
```

Comment:
Shouldn't the `reader` be in `try (...)` ?

Reply:
Good question. The reader is created outside the method, usually from a stream (which is created inside a try-with-resources).

But it doesn't hurt to make sure it's closed even here.